File: selected_coding_interview/codes/cpp/lc_226_invert_binary_tree/lc_226_invert_binary_tree_s1.cpp
/*
 * File: lc_226_invert_binary_tree_s1.cpp
 * Created Time: 2025-12-30
 * Author: krahets
 */

#include "../include/include.hpp"

// ===== Solution Code =====
class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        if (root == nullptr) return nullptr;
        TreeNode* tmp = root->left;
        root->left = invertTree(root->right);
        root->right = invertTree(tmp);
        return root;
    }
};

int main() {
    // ======= Test Case =======
    TreeNode* root = vectorToTree({4, 2, 7, 1, 3, 6, 9});

    // ====== Driver Code ======
    Solution* slt = new Solution();
    TreeNode* res = slt->invertTree(root);
    PrintUtil::printTree(res);

    return 0;
}

File: selected_coding_interview/codes/java/lc_226_invert_binary_tree/lc_226_invert_binary_tree_s1.java
/*
* File: lc_226_invert_binary_tree_s1.java
* Created Time: 2025-12-30
* Author: krahets
*/

package lc_226_invert_binary_tree;

import include.*;
import java.util.*;

// ===== Solution Code =====
    class Solution {
        public TreeNode invertTree(TreeNode root) {
            if (root == null) return null;
            TreeNode tmp = root.left;
            root.left = invertTree(root.right);
            root.right = invertTree(tmp);
            return root;
        }
    }

public class lc_226_invert_binary_tree_s1 {
    public static void main(String[] args) {
        // ======= Test Case =======
        // Test case 1: Basic binary tree
        TreeNode root = TreeNode.arrToTree(new Integer[]{3, 9, 20, null, null, 15, 7});

        // ====== Driver Code ======
        Solution slt = new Solution();
        var result = slt.invertTree(root);
        System.out.println(result);

    }
}
